src/admin_access_token.py

```diff
@@ -66,27 +83,41 @@ def get(self, container: ops.Container) -> typing.Optional[str]:
                     logger.exception("Failed to get secret id %s: %s", secret_id, str(exc))
                     del peer_relation.data[self._app][SECRET_ID]
                     return None
-            else:
-                # There is Secrets support but none was created
-                # So lets create the user and store its token in the secret
-                admin_user = synapse.create_admin_user(container)
-                if not admin_user:
-                    return None
-                logger.debug("Adding secret")
-                secret = self._app.add_secret({SECRET_KEY: admin_user.access_token})
-                peer_relation.data[self._app].update({SECRET_ID: secret.id})
-                admin_access_token = admin_user.access_token
         else:
-            # There is no Secrets support and none relation data was created
-            # So lets create the user and store its token in the peer relation
             secret_value = peer_relation.data[self._app].get(SECRET_KEY)
             if secret_value:
                 admin_access_token = secret_value
-            else:
-                admin_user = synapse.create_admin_user(container)
-                if not admin_user:
-                    return None
-                logger.debug("Adding peer data")
-                peer_relation.data[self._app].update({SECRET_KEY: admin_user.access_token})
-                admin_access_token = admin_user.access_token
+        return admin_access_token
+
+    def _renew_token(self, container: ops.Container) -> typing.Optional[str]:
+        """Create/Renew an admin access token and put it in the peer relation.
+
+        Args:
+            container: Workload container.
+
+        Returns:
+            admin access token or None if there was an error.
+
+        Raise:
+            PeerRelationNotFoundError: if the peer relation does not exist.
+        """
+        peer_relation = self._model.get_relation(PEER_RELATION_NAME)
+        if not peer_relation:
+            logger.error(
+                "Failed to get admin access token: no peer relation %s found", PEER_RELATION_NAME
+            )
+            raise PeerRelationNotFoundError()
+        admin_user = synapse.create_admin_user(container)
+        if not admin_user:
+            logger.error("Error creating admin user to get admin access token")
+            return None
```

Comment:
How about raising an exception here too?

Reply:
I believe there is code in Synapse that depends on that, I would rather not do it in this PR